game-state-service

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python3
+
+import json
+import traceback
+
+from gi.repository import Gio
+from gi.repository import GLib
+
+
+class GameState(object):
+
+    _DEFERRED_TIME = 250
+
+    def __init__(self):
+        self._state = {}
+        self._state_save_deferred_id = None
+        self._state_from_file()
+
+    def _state_get_file(self):
+        path = GLib.build_filenamev([GLib.get_user_data_dir(), 'state.json'])
+        return Gio.File.new_for_path(path)
+
+    def _state_from_file(self):
+        state_file = self._state_get_file()
+        try:
+            success, state_data, tag = state_file.load_contents(None)
+            self._state = json.loads(state_data.decode())
+        except (GLib.Error, json.decoder.JSONDecodeError):
+            traceback.print_exc()
+
+    def _state_to_file(self):
+        state_data = json.dumps(self._state,
+                                sort_keys=True,
+                                indent=2,
+                                separators=(',', ': '))
+        state_file = self._state_get_file()
+        try:
+            state_file.replace_contents(
+                bytes(state_data.encode()),
+                None,
+                False,
+                0,
+                None)
+        except GLib.Error:
+            traceback.print_exc()
+        finally:
+            self._state_save_deferred_id = None
+
+    def _state_save_deferred(self):
+        # piggy back to the existing deferred timeout
+        if self._state_save_deferred_id is not None:
+            return
+        self._state_save_deferred_id = GLib.timeout_add(
+            self._DEFERRED_TIME,
+            self._on_state_save_deferred)
+
+    def _on_state_save_deferred(self):
+        self._state_to_file()
+        self._state_save_deferred_id = None
+        return GLib.SOURCE_REMOVE
+
+    def get(self, key):
+        return self._state[key]
+
+    def set(self, key, value):
+        self._state[key] = value
+        self._state_save_deferred()
+
+
+class GameStateService(object):
+
+    _DBUS_NAME = 'com.endlessm.GameStateService'
+    _DBUS_IFACE = 'com.endlessm.GameStateService'
+    _DBUS_PATH = '/com/endlessm/GameStateService'
+    _DBUS_KEY_ERROR = 'com.endlessm.GameStateService.KeyError'
+    _DBUS_KEY_MESSAGE = 'key %s does not exist'
+    _DBUS_SIGNATURE_ERROR = 'com.endlessm.GameStateService.SignatureError'
+    _DBUS_SIGNATURE_MESSAGE = 'signature %s does not match key %s'
```

Comment:
It's a bit strange to have these format strings up here away from the format parameters, when they're only used in one place anyway. I'll leave that up to you whether you want to change that or keep all the strings in the same place.